docker_watcher: Move docker_watcher into pkg/workloads/containerd

tgraf opened the pull request:
* Unties the docker watcher from the daemon and isolates it in a new
  workloads package. This prepares to support other runtimes in the
  future.

* Unties IPAM from the daemon and moves it into `pkg/ipam`. Only the
  the API handler layer remains in the daemon.

Fixes: #1595

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on daemon/config.go:
```diff
@@ -43,22 +42,18 @@ const (
-	Device         string     // Receive device
-	HostV4Addr     net.IP     // Host v4 address of the snooping device
-	HostV6Addr     net.IP     // Host v6 address of the snooping device
-	DockerEndpoint string     // Docker endpoint
```
What was `DockerEndpoint` used for?

tgraf replied:
This is now passed in as via `func Init(endpoint string) error {`

ianvernon commented on daemon/endpoint.go:
```diff
@@ -647,3 +609,65 @@ func (h *putEndpointIDLabels) Handle(params PutEndpointIDLabelsParams) middlewar
+		ep.Regenerate(d)
+
+		// FIXME: Does this rebuild epID twice?
+		d.TriggerPolicyUpdates([]policy.NumericIdentity{identity.ID})
```
`d.TriggerPolicyUpdates` triggers endpoint regeneration (see code snippet below from daemon's `TriggerPolicyUpdates`), so I don't think you need `ep.Regenerate(d)` on line 666:
```
for k := range Endpoints {
		go func(ep *endpoint.Endpoint, wg *sync.WaitGroup) {
			policyChanges, err := ep.TriggerPolicyUpdates(owner)
			if err != nil {
				log.Warningf("Error while handling policy updates for endpoint %s", err)
				ep.LogStatus(endpoint.Policy, endpoint.Failure, err.Error())
			} else {
				ep.LogStatusOK(endpoint.Policy, "Policy regenerated")
			}
			if policyChanges {
				<-ep.Regenerate(owner)
			}
			wg.Done()
		}(Endpoints[k], &wg)
	}
```

aanm replied:
I think we should tackle on which occasions we need an endpoint regeneration and on which occasions we are regenerating unnecessary. For now are regenerating away more than we need to but it's something we clearly need to fix.

tgraf replied:
I agree with Ian, this PR is leaving this code unmodified. I just moved it over. I'd rather tackle generation optimization in a separate PR.

ianvernon replied:
Ah, OK - we can address this later, then.

ianvernon commented on daemon/state.go:
```diff
@@ -23,14 +23,28 @@ import (
-	ctx "golang.org/x/net/context"
 )
 
+// cleanUpDockerDanglingEndpoints cleans all endpoints that are dandling by
```
dangling

tgraf replied:
fixed

aanm commented on daemon/main.go:
```diff
@@ -90,6 +91,7 @@ var (
 	v6ServicePrefix       string
 	k8sAPIServer          string
 	k8sKubeConfigPath     string
+	dockerEndpoint        string
```
Don't you want to move this to `pkg/containerd`?

tgraf replied:
This is just a temporary storage place to pass it into the `Init()` function later.

aanm commented on pkg/ipam/ipam_test.go:
```diff
@@ -0,0 +1,62 @@
+var _ = Suite(&IPAMSuite{})
+
+func (s *IPAMSuite) TestLock(c *C) {
+	Init()
```
```
err := Init()
c.Assert(err, IsNil)
```

tgraf replied:
fixed

aanm commented on pkg/ipam/types.go:
```diff
@@ -12,24 +12,19 @@
+	IPAMConfig     allocator.IPAMConfig
+	IPv6Allocator  *ipallocator.Range
+	IPv4Allocator  *ipallocator.Range
+	AllocatorMutex lock.RWMutex
```
- Can't the mutex be private?

- Can you put the allocator in the line(s) above the parameters that is protecting?

tgraf replied:
made it private. the mutex protects the entire struct. added a comment.

aanm commented on pkg/workloads/containerd/client.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+// InitMock initializes the containerd packged with a mocked runtime
+func InitMock() error {
```
Does this (L65-L138) needs to be in here? Can't this be in `_test.go`?

tgraf replied:
I don't mind it being in `client.go`. Keeps the related code together.

aanm replied:
But you could have the `client_test.go`

tgraf replied:
How? The code will no longer be included.

aanm replied:
I thought InitMock was not being used outside the `containerd` pkg

aanm commented on pkg/workloads/containerd/watcher.go:
```diff
@@ -49,36 +50,63 @@ const (
 	var wg sync.WaitGroup
 
-	cList, err := d.dockerClient.ContainerList(ctx.Background(), dTypes.ContainerListOptions{All: false})
+	// FIXME: Must be synchronize with event handler
```
Can you clarify this a little more?

tgraf replied:
docker events and the manual sync can overlap and we could readd a container after we have removed it based on the event if the list occurred before the delete event was sent out. I'll file a github issue instead.

aanm reviewed (changes requested):
So we can't support 2 container runtimes at the same time? I'm asking based on the overall code + the cilium status function where we are only checking for containerd status.

tgraf committed:
docker_watcher: Move docker_watcher into pkg/workloads/containerd

* Unties the docker watcher from the daemon and isolates it in a new
  workloads package. This prepares to support other runtimes in the
  future.

* Unties IPAM from the daemon and moves it into `pkg/ipam`. Only the
  the API handler layer remains in the daemon.

Fixes: #1595

Signed-off-by: Thomas Graf <thomas@cilium.io>
